File: 101. Symmetric Tree.py
from typing import Optional, List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree_from_list(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        current = queue.popleft()

        # Ліва дитина
        if i < len(values) and values[i] is not None:
            current.left = TreeNode(values[i])
            queue.append(current.left)
        i += 1

        # Права дитина
        if i < len(values) and values[i] is not None:
            current.right = TreeNode(values[i])
            queue.append(current.right)
        i += 1

    return root


def print_tree_pretty(node: Optional[TreeNode], prefix: str = "", is_left: bool = True):
    if node is not None:
        print_tree_pretty(node.right, prefix + ("│   " if is_left else "    "), False)
        print(prefix + ("└── " if is_left else "┌── ") + str(node.val))
        print_tree_pretty(node.left, prefix + ("    " if is_left else "│   "), True)


values_list = [1, 2, 2, 3, 4, 4, 3]
tree = build_tree_from_list(values_list)
print_tree_pretty(tree)


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def is_mirror(t1, t2):
            if not t1 and not t2:
                return True
            if not t1 or not t2:
                return False
            return (t1.val == t2.val and
                    is_mirror(t1.left, t2.right) and
                    is_mirror(t1.right, t2.left))

        return is_mirror(root.left, root.right) if root else True

sol = Solution()
print(sol.isSymmetric(tree))

File: 101. Symmetric Tree.js
function TreeNode(val, left, right) {
   this.val = (val===undefined ? 0 : val)
   this.left = (left===undefined ? null : left)
   this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */

function build_Tree_FromList(array) {
  if (!array || array.length === 0) return null;

  const root = new TreeNode(array[0]);
  const queue = [root];
  let i = 1;

  while (queue.length > 0 && i < array.length) {
    const current = queue.shift();

    if (i < array.length) {
      if (array[i] != null) {
        current.left = new TreeNode(array[i]);
        queue.push(current.left);
      }
      i++;
    }

    if (i < array.length) {
      if (array[i] != null) {
        current.right = new TreeNode(array[i]);
        queue.push(current.right);
      }
      i++;
    }
  }

  return root;

var isSymmetric = function(root) {
    function isMirror (t1, t2) {
      if (!t1 && !t2) {
        return true;
      }
      if (!t1 || !t2) {
        return false;
      }
      return (t1.val === t2.val && isMirror(t1.left, t2.right) && isMirror(t1.right, t2.left));
    }
    if (root != null) {
      return isMirror(root.left, root.right);
    }
    return true;
};

const tree = build_Tree_FromList([1, 2, 2, 3, 4, 4, 3]);
console.log(isSymmetric(tree));